src/js/index.js

```diff
@@ -56,38 +59,90 @@ RequestUtil.json = function (options = {}) {
         </Provider>),
         domElement);
     } else {
-      setTimeout(function () {
-        let routes = RouterUtil.buildRoutes(appRoutes.getRoutes());
-        let router = Router.run(routes, function (Handler, state) {
-          Config.setOverrides(state.query);
-          ReactDOM.render(
-            (<Provider store={PluginSDK.Store}>
-              <Handler state={state} />
-            </Provider>),
-            domElement);
+      let loadedResources = {
+        mesos: false
+      };
+      let mesosEvents = [MESOS_SUMMARY_CHANGE, MESOS_SUMMARY_REQUEST_ERROR];
+
+      // Responsible for checking if application is ready to render
+      function delayOrLoadApplication() {
+        // Check if we've loaded all resources first
+        let resources = Object.keys(loadedResources);
+        let allLoaded = resources.reduce(function (memo, resource) {
+          if (loadedResources[resource] === false) {
+            memo = false;
+          }
+
+          return memo;
+        }, true);
+
+        if (allLoaded === false) {
+          return;
+        }
+
+        // Next - check to see if enough time has elapsed before rendering
+        let timeSpentLoading = Date.now() - global.getPageLoadedTime();
+        let msLeftOfDelay = Config.applicationRenderDelay - timeSpentLoading;
+
+        if (msLeftOfDelay <= 0) {
+          renderApplicationToDOM();
+        } else {
+          setTimeout(delayOrLoadApplication, msLeftOfDelay);
+        }
+      }
+
+      // Let's make sure we get Mesos Summary data before we render app
+      // Mesos may unreachable, so we will render even on request failure
+      function onMesosSummaryChange() {
+        // Keep polling until the system attaches another listener to summary
+        function keepPollingAlive() {
+          if (MesosSummaryStore.listeners(MESOS_SUMMARY_CHANGE).length > 1) {
+            MesosSummaryStore.removeChangeListener(
+              MESOS_SUMMARY_CHANGE, keepPollingAlive
+            );
+          }
+        }
+        MesosSummaryStore.addChangeListener(
+          MESOS_SUMMARY_CHANGE, keepPollingAlive
+        );
+
+        mesosEvents.forEach(function (event) {
+          MesosSummaryStore.removeChangeListener(event, onMesosSummaryChange);
         });
-
-        PluginSDK.Hooks.doAction('applicationRouter', router);
+        loadedResources.mesos = true;
+        delayOrLoadApplication();
+      }
+
+      // Mesos
+      MesosSummaryStore.init();
+      mesosEvents.forEach(function (event) {
+        MesosSummaryStore.addChangeListener(event, onMesosSummaryChange);
       });
+
+      function renderApplicationToDOM() {
+        setTimeout(function () {
```

Comment:
#nit what do you think about wrapping the call to `renderApplicationToDOM` in L88 with this `setTimeout` instead? Keeps this function clean.